### PR TITLE
[HttpFoundation] changed MERGE queries

### LINKED ISSUE #17284: PdoSessionHandler with SQLSRV PDO driver and SQL server 2008 not work
The PdoSessionHandler not work with SQL Server 2008 or greater.

The query [MERGE](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php#L674) use twice every param.

The error returned by the driver is
```
SQLSTATE[07002]: [Microsoft][ODBC Driver 11 for SQL Server] COUNT field incorrect or syntax error
```

How to use the PDO session handler ?

### PR DESCRIPTION
**Pull request by hjkl**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17284 |
| License | MIT |
| Doc PR | N/A |

Changed the MERGE queries for Oracle and SQL Server to use question mark parameter markers so they work with emulation disabled or enabled - fixes #17284

### CODE REVIEW

**Comment by xabbuh** [comment]
Imo a proper fix would rather change the queries to not use a named parameter more than once.

**Comment by slootjes** [comment]
I can confirm this fix is working for me.

**Comment by fabpot** [comment]
@Tobion Can you have a look at this one?

**Comment by fabpot** [comment]
@hjkl Have you tested @xabbuh's suggestion instead?

**Comment by Tobion** [comment]
@hjkl please change `getMergeSql` (probably rename it as well) to return the PDO statement directly with the params already bound. Since the params and the sql are now dependent on each other, it should be done in one place.
Apart from that it looks correct to me.

**Comment by Tobion** [comment]
This PR needs a rebase.

**Comment by hjkl** [comment]
@fabpot I haven't attempted @xabbuh's suggestion - I don't have the TSQL expertise to refactor the merge query to only use each parameter once.

@Tobion Done, thanks for the advice.

**Comment by Tobion** [comment]
> Imo a proper fix would rather change the queries to not use a named parameter more than once.

This is exactly what this PR does. NAMED params cannot be used more than once but it's not possible to change the MERGE queries to only use each param once as they need to be repeated. It only works for the MySQL for example because mysql supports the ` VALUES($this->dataCol)` syntax to reuse values. Otherwise we would also be forced into using numeric values several times. For reference, if you disable emulated prepares in MySQL and use the same named param more than once you get a `SQLSTATE[HY093]: Invalid parameter number` error as I just tried.

:+1: good to merge
Status: Reviewed

**Comment by fabpot** [comment]
Thank you @hjkl.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 1]
```diff
@@ -653,32 +647,65 @@ private function getSelectSql()
...
+            if ('sqlsrv' === $this->driver || 'oci' === $this->driver) {
+                $mergeStmt->bindParam(1, $sessionId, \PDO::PARAM_STR);
+                $mergeStmt->bindParam(2, $sessionId, \PDO::PARAM_STR);
+                $mergeStmt->bindParam(3, $data, \PDO::PARAM_LOB);
+                $mergeStmt->bindParam(4, $maxlifetime, \PDO::PARAM_INT);
+                $mergeStmt->bindValue(5, time(), \PDO::PARAM_INT);
+                $mergeStmt->bindParam(6, $data, \PDO::PARAM_LOB);
+                $mergeStmt->bindParam(7, $maxlifetime, \PDO::PARAM_INT);
+                $mergeStmt->bindValue(8, time(), \PDO::PARAM_INT);
+            }
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: @hjkl, could you please submit another PR fixing this minor CS issue? Thank you in advance.

Comment (hjkl) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: @phansys sure, I can't see the coding standards issue though, could you please clarify for me?

Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: Sure @hjkl, I refer to [PSR-2 5.1](http://www.php-fig.org/psr/psr-2/#5-1-if-elseif-else):
```php
} else {
```
Thank you.

**Pull request merged**